frontend/include/chpl/resolution/resolution-types.h

```diff
@@ -866,6 +877,147 @@ class TypedFnSignature {
   /// \endcond DO_NOT_DOCUMENT
 };
 
+enum CandidateFailureReason {
+  /* Cannot pass an actual to one of the candidate's formals. */
+  FAIL_CANNOT_PASS,
+  /* Not a valid formal-actual mapping for this candidate. */
+  FAIL_FORMAL_ACTUAL_MISMATCH,
+  /* The wrong number of varargs were given to the function. */
+  FAIL_VARARG_MISMATCH,
+  /* The where clause returned 'false'. */
+  FAIL_WHERE_CLAUSE,
+  /* A parenful call to a parenless function or vice versa. */
+  FAIL_PARENLESS_MISMATCH,
+  /* Some other, generic reason. */
+  FAIL_CANDIDATE_OTHER,
+};
+
+enum PassingFailureReason {
+  /* Incompatible nilability (e.g. nilable to non-nilable formal) */
+  FAIL_INCOMPATIBLE_NILABILITY,
+  /* Incompatible management (e.g. borrowed to owned formal) */
+  FAIL_INCOMPATIBLE_MGMT,
+  /* Incompatible manager (e.g. owned to shared formal) */
+  FAIL_INCOMPATIBLE_MGR,
+  /* The actual is not a subtype of the formal, but had to be. */
+  FAIL_EXPECTED_SUBTYPE,
+  /* Tuple size doesn't match */
+  FAIL_INCOMPATIBLE_TUPLE_SIZE,
+  /* Tuple star-ness doesn't match */
+  FAIL_INCOMPATIBLE_TUPLE_STAR,
+  /* A conversion was needed but is not possible. */
+  FAIL_CANNOT_CONVERT,
+  /* An instantiation was needed but is not possible. */
+  FAIL_CANNOT_INSTANTIATE,
+  /* A type was used as an argument to a value, or the other way around. */
+  FAIL_TYPE_VS_NONTYPE,
+  /* A param value was expected, but a non-param value was given. */
+  FAIL_NOT_PARAM,
+  /* One param value was expected, another was given. */
+  FAIL_MISMATCHED_PARAM,
+  /* An unestablished actual type was given to the function. */
+  FAIL_UNKNOWN_ACTUAL_TYPE,
+  /* An unestablished formal type in the function. */
+  FAIL_UNKNOWN_FORMAL_TYPE,
+  /* A generic type was given as an argument to a non-type formal. */
+  FAIL_GENERIC_TO_NONTYPE,
+  /* A type was expected to be the exact match of the formal, but wasn't. */
+  FAIL_NOT_EXACT_MATCH,
+  /* Some other, generic reason. */
+  FAIL_FORMAL_OTHER,
+};
+
+class ApplicabilityResult {
+ private:
+  ID idForErr_;
+  const TypedFnSignature* initialForErr_;
+  const TypedFnSignature* candidate_;
+  CandidateFailureReason candidateReason_;
+  PassingFailureReason formalReason_;
+  int formalIdx_;
+
+  ApplicabilityResult(ID idForErr,
+                      const TypedFnSignature* initialForErr,
+                      const TypedFnSignature* candidate,
+                      CandidateFailureReason candidateReason,
+                      PassingFailureReason formalReason,
+                      int formalIdx) :
+    idForErr_(std::move(idForErr)), initialForErr_(initialForErr), candidate_(candidate),
+    candidateReason_(candidateReason), formalReason_(formalReason),
+    formalIdx_(formalIdx) {
+    CHPL_ASSERT(!candidate_ || (formalIdx_ == -1 && formalReason_ == FAIL_FORMAL_OTHER));
```

Comment:
What is the purpose of this assert? I'm reading it as "either there was no candidate, or the candidate is invalid due to a generic passing failure", and I don't understand why we would want that.

Reply:
The purpose of the assert is to ensure that if the result was a _success_ (candidate is set), then there's no information about what went wrong, because nothing went wrong. Note that "~p or q` is the as `p implies q`, so the above could be read as "resolution succeeding implies error information is all dummy / templated".